### PR TITLE
IEEE 802.15.4 driver API changes to match OT certification requirements

### PR DESCRIPTION
**Pull request by tbursztyka**

### CODE REVIEW

File: drivers/ieee802154/ieee802154_cc1200.c [hunk 1]
```diff
@@ -435,13 +435,10 @@ static inline bool read_rxfifo_content(struct cc1200_context *ctx,
 	return true;
 }
 
-static inline bool verify_crc(struct cc1200_context *ctx, struct net_pkt *pkt)
+static inline bool verify_crc(struct cc1200_context *ctx,
```
[the remark this answers is not in the document]
Comment (tbursztyka) [reply drivers/ieee802154/ieee802154_cc1200.c]: indeed, and actually context is unnecessary here.

File: drivers/ieee802154/ieee802154_kw41z.c [hunk 2]
```diff
@@ -530,6 +530,11 @@ static inline void kw41z_rx(struct kw41z_context *kw41z, u8_t pkt_len)
 	net_pkt_set_ieee802154_lqi(pkt, kw41z_convert_lqi(hw_lqi));
 	/* ToDo: get the rssi as well and use net_pkt_set_ieee802154_rssi() */
 
+	if (ieee802154_radio_handle_ack(kw41z->iface, pkt) == NET_OK) {
+		SYS_LOG_DBG("ACK packet handled");
```
Comment (dleach02) [inline drivers/ieee802154/ieee802154_kw41z.c]: Just FYI but the HW is configured to handle the ACKs so they won't be passed up to the driver...

Comment (tbursztyka) [reply drivers/ieee802154/ieee802154_kw41z.c]: oh yes, completely forgot about IEEE802154_HW_TX_RX_ACK capability. Vacations were nice.

**Comment by tbursztyka** [comment]
Ok, I had issues with the FCS commit: basically the current buffer management being what it is (aka "not so smart"), I had to make a very crude hack in net_pkt when allocating tx data. Please see relevant commit.

This will hopefully disappear in a next-to-come PR where I'll switch net_pkt to use variable size memory allocator for data fragments.

**Review by rlubos — APPROVED** [review]
LGTM :+1:

**Comment by tbursztyka** [comment]
I am not super happy about that hack. I might revert to #ifdef in drivers instead. (which would disappear once we get a proper buffer management in the core)

File: include/net/ieee802154_radio.h [hunk 3]
```diff
@@ -96,6 +96,16 @@ struct ieee802154_radio_api {
 	/** Get the available amount of Sub-GHz channels */
 	u16_t (*get_subg_channel_count)(struct device *dev);
 #endif /* CONFIG_NET_L2_IEEE802154_SUB_GHZ */
+
+#ifdef CONFIG_NET_L2_OPENTHREAD
+	/** Run an energy detection scan.
+	 * Note: channel must be set prior to request this function.
+	 */
+	int (*ed_scan)(struct device *dev,
+		       u16_t duration,
```
Comment (jukkar) [inline include/net/ieee802154_radio.h]: I wonder what the duration means here meaning seconds, milliseconds or something else, perhaps document it here?

Comment (dleach02) [reply include/net/ieee802154_radio.h]: According to the header file from Thread:

/**
 * This method begins the energy scan sequence on the radio.
 *
 * @param[in] aInstance      The OpenThread instance structure.
 * @param[in] aScanChannel   The channel to perform the energy scan on.
 * @param[in] aScanDuration  The duration, in milliseconds, for the channel to be scanned.
 *
 * @retval OT_ERROR_NONE             Successfully started scanning the channel.
 * @retval OT_ERROR_NOT_IMPLEMENTED  The radio doesn't support energy scanning.
 */
otError otPlatRadioEnergyScan(otInstance *aInstance, uint8_t aScanChannel, uint16_t aScanDuration);

**Comment by tbursztyka** [comment]
Ok I removed the patch making L2 aware of FCS, the hack was way too crappy and this will be fully solved in net buffer management change. currently OT enabled drivers do already what's supposed to be done (aka: passing the FCS to OT's L2) so it's fine.

**Review by jukkar — APPROVED** [review]

**Pull request merged**